lib/spack/spack/fetch_strategy.py

```diff
@@ -158,64 +194,65 @@ class FetchStrategyComposite(object):
     Implements the GoF composite pattern.
     """
     matches = FetchStrategy.matches
-    set_stage = FetchStrategy.set_stage
 
 
+@fetcher
 class URLFetchStrategy(FetchStrategy):
     """FetchStrategy that pulls source code from a URL for an archive,
        checks the archive against a checksum,and decompresses the archive.
     """
-    enabled = True
     required_attributes = ['url']
+    #: Overridden by instances to set the curl executable
+    _curl = None
 
-    def __init__(self, url=None, digest=None, **kwargs):
-        super(URLFetchStrategy, self).__init__()
-
-        # If URL or digest are provided in the kwargs, then prefer
-        # those values.
-        self.url = kwargs.get('url', None)
-        if not self.url:
-            self.url = url
-
-        self.digest = next((kwargs[h] for h in crypto.hashes if h in kwargs),
-                           None)
+    def __init__(self, url, digest=None, **kwargs):
+        self.url = url
+        self.digest = next(
+            (kwargs[h] for h in crypto.hashes if h in kwargs), None
```

Comment:
I think the way I did this was bad (but definitely not the worst thing I did with fetchers -- so much of them is horrible).  There should be a failure here if they provide two digest types... also, we allow them to just pass the hash in without a name... there should probably be an error if they pass a `digest` argument _and_ a hash.  Ugh.

How about:
```python
digests = [] if digest is None else [digest]
digests += [kwargs[h] for h in kwargs if h in crypto.hashes]
if len(digests) > 1:
    raise ValueError('Fetchers can only use one digest')
self.digest = digests[0] if digests else None
```

Reply:
Part of the modifications here were along the lines of being defensive. I didn't want to pile up too many modifications on behavior without knowing why you did some of the things here and while I was modifying the architecture of stages and fetchers.